packer/builder/azure/config.go

```diff
@@ -129,6 +131,20 @@ func newConfig(raws ...interface{}) (*Config, []string, error) {
 		errs = packer.MultiErrorAppend(errs, fmt.Errorf("instance_size is not valid, must be one of: %v", allowedVMSizes))
 	}
 
+	for n := 0; n < len(c.DataDisks); n++ {
+		switch v := c.DataDisks[n].(type) {
+		case string:
```

Comment:
it's not discarding it, it's generating an error. You say
```json
{
  ...
  "data_disks": [
    10,
    "http://blob.core.windows.net/container/sourceimage.vhd"
  ]
}
```
  a number means 'create a new disk' with that many GB
  a url means 'take this image as the source for the disk'

So indeed, for the current design, only numbers and strings are valid items in the array.